wallet-api/src/Ledger/Map/TH.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE UndecidableInstances   #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE DeriveAnyClass         #-}
+{-# LANGUAGE DeriveGeneric          #-}
+{-# LANGUAGE DerivingStrategies     #-}
+{-# LANGUAGE TemplateHaskell        #-}
+{-# LANGUAGE LambdaCase             #-}
+{-# LANGUAGE MonoLocalBinds         #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+-- A map implementation that can be used in on-chain and off-chain code.
+module Ledger.Map.TH(
+    Map
+    , IsEqual
+    , singleton
+    , empty
+    , map
+    , lookup
+    , union
+    , all
+    -- * These
+    , These(..)
+    , these
+    ) where
+
+import           Codec.Serialise.Class        (Serialise)
+import           Data.Aeson                   (FromJSON, ToJSON)
+import           Data.Swagger.Internal.Schema (ToSchema)
+import           GHC.Generics                 (Generic)
+import           Language.PlutusTx.Lift       (makeLift)
+import qualified Language.PlutusTx.Prelude    as P
+import           Language.Haskell.TH          (Q, TExp)
+import           Ledger.These.TH              (These(..), these)
+import           Prelude                      hiding (all, lookup, map, negate)
+
+
+-- | A 'Map' of key-value pairs. 
+data Map k v = Map { unMap :: [(k, v)] }
+    deriving (Show)
+    deriving stock (Generic)
+    deriving anyclass (ToSchema, ToJSON, FromJSON, Serialise)
+
+makeLift ''Map
+
+-- | Apply a function to the values of a 'Map'.
+map :: Q (TExp ((v -> w) -> Map k v -> Map k w))
+map = [|| 
+    let
+        map :: forall k v w. (v -> w) -> Map k v -> Map k w
+        map f (Map mp) =
+            let 
+                go :: [(k, v)] -> [(k, w)]
+                go [] = []
+                go ((c, i):xs') = (c, f i) : go xs'
+            in Map (go mp)
+    in
+        map
+    ||]
+
+-- | Compare two 'k's for equality.
+type IsEqual k = k -> k -> Bool

```

Comment:
I suggest we instead require a `type Comparison k = k -> k -> Ordering`. That's more power than we need here, but that way when we change to using binary search trees we won't have to change all the clients.

Reply:
I agree in principle, but then we also need a new builtin `geqByteString` if we want to use this in on-chain code.

Reply:
Okay, makes sense for now. We'll need that builtin eventually, though, so let's add it.